File: backend/nodejs/apps/tests/helpers/fakeClusterRedis.ts
/**
 * In-memory Redis Cluster double for unit tests (R5, R17).
 *
 * A minimal hand-rolled Redis command surface, enforcing the one rule real
 * Redis Cluster / AWS MemoryDB enforce that the plain `FakeRedis` in
 * `mock-ioredis-global.ts` does not: a command touching more than one key
 * must have every key land in the same hash slot, or the cluster node
 * itself rejects it with CROSSSLOT. `cluster-key-slot` (already a
 * transitive `ioredis` dependency, promoted to an explicit devDependency
 * for this file) computes the slot exactly the way `ioredis.Cluster` and
 * `redis.crc.key_slot` (Python) do, including `{hashtag}` support.
 *
 * This is what makes a unit test fail if shared Node.js code regresses back
 * to a multi-key command that only breaks in production against a real
 * cluster or MemoryDB endpoint.
 *
 * Two instances built with the same backing `FakeClusterStore` (via the
 * `store` constructor option) behave like two client connections to one
 * real server -- mirrors `IRedisConnectionProvider.getClient()` and
 * `.createClient()` handing out different objects for the same endpoint.
 *
 * Usage:
 * ```ts
 * const fake = new FakeClusterRedis();
 * await fake.set('a', '1');
 * await fake.set('b', '2');
 * await fake.mget(['a', 'b']); // rejects with a CrossSlotError
 * ```
 */
import { EventEmitter } from 'events';

import calcSlot from 'cluster-key-slot';

export class CrossSlotError extends Error {
  constructor(keys: string[], slots: number[]) {
    super(
      `CROSSSLOT Keys in request don't hash to the same slot: ${JSON.stringify(
        keys,
      )} -> slots [${slots.join(', ')}]`,
    );
    this.name = 'CrossSlotError';
  }
}

function assertSameSlot(keys: string[]): void {
  if (keys.length <= 1) {
    return;
  }
  const slots = Array.from(new Set(keys.map((key) => calcSlot(key))));
  if (slots.length > 1) {
    throw new CrossSlotError(keys, slots);
  }
}

/** Backing dataset, shareable across several `FakeClusterRedis` instances. */
export class FakeClusterStore {
  readonly strings = new Map<string, string>();
  readonly hashes = new Map<string, Map<string, string>>();
  readonly scripts = new Map<string, string>();
}

interface FakePipelineCommand {
  name: string;
  args: unknown[];
}

class FakeClusterPipeline {
  private readonly commands: FakePipelineCommand[] = [];

  constructor(private readonly client: FakeClusterRedis) {}

  private queue(name: string, ...args: unknown[]): this {
    this.commands.push({ name, args });
    return this;
  }

  get(...args: unknown[]): this {
    return this.queue('get', ...args);
  }

  set(...args: unknown[]): this {
    return this.queue('set', ...args);
  }

  del(...args: unknown[]): this {
    return this.queue('del', ...args);
  }

  hget(...args: unknown[]): this {
    return this.queue('hget', ...args);
  }

  hset(...args: unknown[]): this {
    return this.queue('hset', ...args);
  }

  /**
   * Each queued command is dispatched independently -- exactly like a real
   * `ClusterPipeline`, which routes every command to the node owning its
   * own key(s) rather than requiring the whole batch to share a slot.
   */
  async exec(): Promise<Array<[Error | null, unknown]>> {
    const results: Array<[Error | null, unknown]> = [];
    for (const { name, args } of this.commands) {
      try {
        const value = await (this.client as any)[name](...args);
        results.push([null, value]);
      } catch (error) {
        results.push([error as Error, null]);
      }
    }
    return results;
  }
}

/** Same command surface, without the `[error, value]` tuple wrapping --
 * matches `ioredis`'s non-transactional pipeline result shape used by
 * callers that pass `{ transaction: false }`-style plain arrays. */
class FakeClusterMulti extends FakeClusterPipeline {}

export class FakeClusterRedis extends EventEmitter {
  status = 'ready';

  private readonly store: FakeClusterStore;

  constructor(store: FakeClusterStore = new FakeClusterStore()) {
    super();
    this.store = store;
    process.nextTick(() => {
      this.emit('connect');
      this.emit('ready');
    });
  }

  /** Pass to `new FakeClusterRedis(store)` so both wrap the same dataset. */
  get sharedStore(): FakeClusterStore {
    return this.store;
  }

  keySlot(key: string): number {
    return calcSlot(key);
  }

  async get(key: string): Promise<string | null> {
    return this.store.strings.get(key) ?? null;
  }

  async set(key: string, value: string): Promise<'OK'> {
    this.store.strings.set(key, value);
    return 'OK';
  }

  async del(...keys: string[]): Promise<number> {
    assertSameSlot(keys);
    let deleted = 0;
    for (const key of keys) {
      if (this.store.strings.delete(key) || this.store.hashes.delete(key)) {
        deleted += 1;
      }
    }
    return deleted;
  }

  async mget(keys: string[]): Promise<Array<string | null>> {
    assertSameSlot(keys);
    return keys.map((key) => this.store.strings.get(key) ?? null);
  }

  async exists(...keys: string[]): Promise<number> {
    assertSameSlot(keys);
    return keys.filter(
      (key) => this.store.strings.has(key) || this.store.hashes.has(key),
    ).length;
  }

  async incr(key: string): Promise<number> {
    const current = Number(this.store.strings.get(key) ?? '0') + 1;
    this.store.strings.set(key, String(current));
    return current;
  }

  async expire(_key: string, _seconds: number): Promise<number> {
    return 1;
  }

  async hget(key: string, field: string): Promise<string | null> {
    return this.store.hashes.get(key)?.get(field) ?? null;
  }

  async hset(key: string, field: string, value: string): Promise<number> {
    let hash = this.store.hashes.get(key);
    if (!hash) {
      hash = new Map();
      this.store.hashes.set(key, hash);
    }
    const isNew = !hash.has(field);
    hash.set(field, value);
    return isNew ? 1 : 0;
  }

  async hdel(key: string, field: string): Promise<number> {
    return this.store.hashes.get(key)?.delete(field) ? 1 : 0;
  }

  /** Single-node SCAN (R17): sufficient for exercising `scanKeys` fan-out
   * logic in tests, which only needs the returned keys to be correct, not
   * genuine multi-node cursor semantics. */
  async scan(
    cursor: string,
    _match: string,
    pattern: string,
  ): Promise<[string, string[]]> {
    if (cursor !== '0') {
      return ['0', []];
    }
    const regex = new RegExp(`^${pattern.replace(/\*/g, '.*')}$`);
    const keys = [...this.store.strings.keys(), ...this.store.hashes.keys()].filter(
      (key) => regex.test(key),
    );
    return ['0', keys];
  }

  async script(action: string, body: string): Promise<string> {
    if (action.toUpperCase() !== 'LOAD') {
      throw new Error(`FakeClusterRedis.script only supports LOAD, got ${action}`);
    }
    const sha = `sha-${this.store.scripts.size}-${body.length}`;
    this.store.scripts.set(sha, body);
    return sha;
  }

  async eval(_script: string, numkeys: number, ...rest: unknown[]): Promise<unknown> {
    assertSameSlot(rest.slice(0, numkeys) as string[]);
    return null;
  }

  async evalsha(sha: string, numkeys: number, ...rest: unknown[]): Promise<unknown> {
    assertSameSlot(rest.slice(0, numkeys) as string[]);
    if (!this.store.scripts.has(sha)) {
      const error = new Error(`NOSCRIPT No matching script. Please use EVAL.`);
      error.name = 'NOSCRIPT';
      throw error;
    }
    return null;
  }

  pipeline(_options?: { transaction?: boolean }): FakeClusterPipeline {
    return new FakeClusterPipeline(this);
  }

  multi(_options?: { transaction?: boolean }): FakeClusterMulti {
    return new FakeClusterMulti(this);
  }

  async ping(): Promise<'PONG'> {
    return 'PONG';
  }

  async quit(): Promise<'OK'> {
    return 'OK';
  }

  async disconnect(): Promise<void> {}

  nodes(_role?: 'master' | 'all' | 'slave'): FakeClusterRedis[] {
    return [this];
  }

  defineCommand(name: string, _definition?: unknown): void {
    (this as any)[name] = async () => null;
  }
}

File: backend/python/tests/support/fake_cluster_redis.py
"""In-memory Redis Cluster double for unit tests (R5, R17).

Wraps a single `fakeredis.aioredis.FakeRedis` instance and enforces the one
Redis Cluster rule that `fakeredis` does not: multi-key commands whose keys
land in different hash slots must raise, exactly as `redis-py`'s
`RedisCluster` does for `MGET`/`EVAL`/`EVALSHA`. This is what makes a unit
test fail if shared code (`RetryManager`, `AccessibleRecordsCache`,
`DistributedConcurrencyManager`, ...) regresses back to a multi-key command
that only breaks in production against a real cluster or AWS MemoryDB.

`DEL`/`EXISTS` are deliberately NOT slot-checked here: `redis-py`'s
`RedisClusterPipeline`/command layer already splits those per slot
internally, so they are safe today (see the compatibility matrix in the
MemoryDB readiness plan) -- only `MGET`, `EVAL`, and `EVALSHA` are genuinely
atomic-and-unsplit in a way that raises CROSSSLOT.

Usage::

    fake = FakeClusterRedis()
    manager = RetryManager(logger=..., redis_client=fake)
    await manager.initialize()
    await manager.has_pending_retries(["a", "b"])  # ids may span slots
"""
from __future__ import annotations

from typing import Any

import pytest
from redis.crc import key_slot
from redis.exceptions import ClusterCrossSlotError

# `importorskip`, not a plain `import`: `fakeredis` lives in the optional
# `dev` extra, and every importer of this helper had its own
# `pytest.importorskip("fakeredis.aioredis")` guard placed *below* the line
# that imports this module -- so without the extra installed, collection
# raised ModuleNotFoundError instead of skipping. Guarding here covers every
# importer at once, including the ones that import this lazily inside a test
# body, and any added later.
fakeredis = pytest.importorskip("fakeredis")

__all__ = ["FakeClusterRedis"]


def _to_bytes(key: Any) -> bytes:  # noqa: ANN401 - mirrors redis-py's own key coercion
    return key.encode() if isinstance(key, str) else key


def _assert_same_slot(keys: "list[Any]") -> None:
    if len(keys) <= 1:
        return
    slots = {key_slot(_to_bytes(k)) for k in keys}
    if len(slots) > 1:
        raise ClusterCrossSlotError(
            f"Command spans multiple hash slots: {keys!r} -> slots {sorted(slots)}"
        )


class FakeClusterRedis:
    """`redis.asyncio.Redis`-shaped double that raises CROSSSLOT like a real
    Redis Cluster / MemoryDB deployment would, backed by one in-process
    `fakeredis` server (single-node SCAN and all, since only the multi-key
    command rules -- not sharding -- are what shared code must respect)."""

    def __init__(
        self,
        *,
        server: "fakeredis.FakeServer | None" = None,
        version: tuple[int, ...] = (7,),
        decode_responses: bool = True,
    ) -> None:
        self._server = server or fakeredis.FakeServer(version=version)
        self._redis = fakeredis.aioredis.FakeRedis(
            server=self._server, decode_responses=decode_responses
        )

    @property
    def server(self) -> "fakeredis.FakeServer":
        """Pass to another `FakeClusterRedis(server=...)` so both wrap the
        same in-memory dataset -- mirrors several client objects (`get_client()`,
        `create_client()`, pub/sub) all talking to one real Redis/MemoryDB
        endpoint."""
        return self._server

    async def mget(self, keys: "list[Any]", *args: Any) -> list[Any]:  # noqa: ANN401
        all_keys = list(keys) + list(args) if not isinstance(keys, (str, bytes)) else [keys, *args]
        _assert_same_slot(all_keys)
        return await self._redis.mget(all_keys)

    async def eval(self, script: str, numkeys: int, *keys_and_args: Any) -> Any:  # noqa: ANN401
        _assert_same_slot(list(keys_and_args[:numkeys]))
        return await self._redis.eval(script, numkeys, *keys_and_args)

    async def evalsha(self, sha: str, numkeys: int, *keys_and_args: Any) -> Any:  # noqa: ANN401
        _assert_same_slot(list(keys_and_args[:numkeys]))
        return await self._redis.evalsha(sha, numkeys, *keys_and_args)

    def pipeline(self, transaction: bool = True) -> Any:  # noqa: ANN401
        """Non-transactional pipelines legitimately span slots (a real
        `ClusterPipeline` routes each command to its own node); `transaction=True`
        (WATCH/MULTI/EXEC) does not survive a cluster hop (R3) but nothing in
        this codebase uses it anymore, so it is intentionally left unchecked
        here rather than half-emulated.
        """
        return self._redis.pipeline(transaction=transaction)

    def __getattr__(self, name: str) -> Any:  # noqa: ANN401
        return getattr(self._redis, name)
